### PR TITLE
Fix : User Roles not added to create, update or delete calls

### PR DESCRIPTION
**Pull request by flessard**

### CODE REVIEW

**Comment by gfosco** [comment]
Thanks for attempting to fix this issue.  There are 7 test failures to address before further review.

**Comment by gfosco** [comment]
Down to 4... You can run tests locally with `npm test`

**Comment by flessard** [comment]
Fix this issuse #389 #294

File: spec/ParseRole.spec.js [hunk 1]
```diff
@@ -49,7 +49,7 @@ describe('Parse Role testing', () => {
     }).then((x) => {
       x.set('foo', 'baz');
       // This should fail:
-      return x.save();
+      return x.save({},{sessionToken: ""});
```
Comment (flessard) [inline spec/ParseRole.spec.js]: I change this because, by default the sessionToken is passed. The result of this never fail.

File: src/Auth.js [hunk 2]
```diff
@@ -80,7 +80,7 @@ Auth.prototype.getUserRoles = function() {
     return Promise.resolve(this.userRoles);
   }
   if (this.rolePromise) {
-    return rolePromise;
+    return this.rolePromise;
```
Comment (flovilmart) [inline src/Auth.js]: :+1:

File: src/RestWrite.js [hunk 3]
```diff
@@ -27,6 +27,9 @@ function RestWrite(config, auth, className, query, data, originalData) {
   this.auth = auth;
   this.className = className;
   this.storage = {};
+  this.runOptions = {
+    acl:['*']
```
Comment (flovilmart) [inline src/RestWrite.js]: is it safe to mark as '*' by default?

File: src/rest.js [hunk 4]
```diff
@@ -57,11 +57,18 @@ function del(config, auth, className, objectId) {
     }
     return Promise.resolve({});
   }).then(() => {
+    if (!auth.isMaster) {
+      return auth.getUserRoles();
```
Comment (flovilmart) [inline src/rest.js]: we could add an auth.resolveACL() that would return a proper ACL query object given the current state of the auth.
That would remove the duplication of concatenating with userRoles at both places

File: src/RestWrite.js (hunk 3 above)
Comment (gfosco) [inline src/RestWrite.js]: No, the acl options shouldn't be set by default.  A master key query shouldn't have an acl key set, for example.. See `ExportAdapter.update`

File: src/rest.js (hunk 4 above)
Comment (flessard) [inline src/rest.js]: @flovilmart do you mean the auth.resolveACL will ask the server to get the new state of the user ?  
What does not do getUserRoles by using a cache if already fetched.

Sorry but my English is not perfect :)

Comment (flovilmart) [inline src/rest.js]: I mean a method on auth like:

function queryACLConstraint() {
 // do something
return {}; // the ACL Constraint for the query
}

This inspired by the way we build the ACL options around the app, this way we centralize that logic.

**Comment by flessard** [comment]
@gfosco What is missing on this PR it to be merge?

**Comment by gfosco** [comment]
:+1:

**Pull request merged**